### PR TITLE
Add checksums support

### PR DESCRIPTION
**Pull request by hgeraldino**
Requires new version of sdkman-mongodb-persistence

### CODE REVIEW

File: src/test/resources/features/release_version_checksums.feature [hunk 1]
```diff
@@ -0,0 +1,130 @@
+Feature: Release Version with checksums
+
+  Background:
+    Given the Consumer has a valid Auth Token
+
+  Scenario: Release binary for Linux 64 bit Platform with checksums
+    Given the existing Default PLATFORM_SPECIFIC java Version is 8u121-zulu
+    And the Consumer for candidate java|jmc is making a request
```
Comment (marc0der) [inline src/test/resources/features/release_version_checksums.feature:8]: Should this be java?

File: src/test/resources/features/release_version_checksums.feature [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    """
+    Then the status received is 201 CREATED
+    And java Version 8u131-zulu with URL http://localhost:8080/zulu8.21.0.1-jdk8.0.131-linux_x64.tar.gz was published as LINUX_64
+    And java Version 8u131-zulu on platform LINUX_64 has a checksum "sha1-checksum" using algorithm SHA-1
+    And java Version 8u131-zulu on platform LINUX_64 has a checksum "md5-checksum" using algorithm MD5
+    And java Version 8u131-zulu on platform LINUX_64 has a checksum "sha256-checksum" using algorithm SHA-256
+
+  Scenario: Release multiple multi-platform binaries of the same Version with checksums
+    Given an existing LINUX_64 java Version 8u121-zulu exists
+    And the Consumer for candidate java|jmc is making a request
```
Comment (marc0der) [inline src/test/resources/features/release_version_checksums.feature:32]: As before.

File: src/test/resources/features/release_version_checksums.feature [hunk 3]
```diff
@@ -0,0 +1,130 @@
+Feature: Release Version with checksums
+
+  Background:
+    Given the Consumer has a valid Auth Token
+
+  Scenario: Release binary for Linux 64 bit Platform with checksums
+    Given the existing Default PLATFORM_SPECIFIC java Version is 8u121-zulu
```
Comment (marc0der) [inline src/test/resources/features/release_version_checksums.feature:7]: I would choose a UNIVERSAL candidate, as PLATFORM specific is already covered in the subsequent test.

File: src/test/resources/features/release_version_checksums.feature (hunk 1 above)
Comment (marc0der) [inline src/test/resources/features/release_version_checksums.feature:8]: We should only have one candidate here. Both of these are platform specific candidates. How about Scala, Kotlin or Groovy instead.

File: src/main/scala/io/sdkman/vendor/release/routes/Validation.scala [hunk 4]
```diff
@@ -45,4 +46,35 @@ trait Validation {
       17 >= version.length,
       ApiResponse(400, s"Version length exceeds 17 chars: $version").toJson.compactPrint
     )
+
+  def validateChecksumAlgorithms(checksums: Option[Map[String,String]]): Directive0 =
+    checksums
+      .map { checksum =>
+        val algorithms: Seq[String] = Seq(MD5.id, SHA1.id, SHA224.id, SHA256.id, SHA384.id, SHA512.id)
```
Comment (marc0der) [inline src/main/scala/io/sdkman/vendor/release/routes/Validation.scala:53]: I'd move this outside of the `def` declaration to save it from being evaluated each time the def is called.

File: src/main/scala/io/sdkman/vendor/release/routes/Validation.scala [hunk 5]
```diff
@@ -45,4 +46,35 @@ trait Validation {
...
+
+        val invalidAlgorithm = checksum.keys.find { key =>
+            !algorithms.contains(key)
+        }.getOrElse("")
```
Comment (marc0der) [inline src/main/scala/io/sdkman/vendor/release/routes/Validation.scala:57]: I think this guy will only find the first non-matching algorithm. What if there is more than one?

File: src/main/scala/io/sdkman/vendor/release/routes/Validation.scala [hunk 6]
```diff
@@ -45,4 +46,35 @@ trait Validation {
...
+        validate(
+          check = invalidAlgorithm.isEmpty,
+          ApiResponse(400, s"Invalid checksum algorithm: ${invalidAlgorithm}").toJson.compactPrint)
+
+      } getOrElse pass
+
+  def validateChecksums(checksums: Option[Map[String,String]]): Directive0 =
+    checksums
+      .map { checksum =>
+        var invalidChecksumAlgorithm: String = ""
```
Comment (marc0der) [inline src/main/scala/io/sdkman/vendor/release/routes/Validation.scala:68]: How about we try to do this without mutable state? You could use a similar approach as I outlined above.

File: src/main/scala/io/sdkman/vendor/release/routes/Validation.scala [hunk 7]
```diff
@@ -45,4 +46,35 @@ trait Validation {
...
+            !algorithms.contains(key)
+        }.getOrElse("")
+
+        validate(
+          check = invalidAlgorithm.isEmpty,
+          ApiResponse(400, s"Invalid checksum algorithm: ${invalidAlgorithm}").toJson.compactPrint)
+
+      } getOrElse pass
+
+  def validateChecksums(checksums: Option[Map[String,String]]): Directive0 =
```
Comment (marc0der) [inline src/main/scala/io/sdkman/vendor/release/routes/Validation.scala:65]: This functiona and the previous one feel so similar that we could probably extract a higher order function. wdyt?

File: src/main/scala/io/sdkman/vendor/release/routes/Validation.scala [hunk 8]
```diff
@@ -45,4 +46,35 @@ trait Validation {
...
+    checksums
+      .map { checksum =>
+        var invalidChecksumAlgorithm: String = ""
+
+        for((k, v) <- checksum) {
+          if(v == null || v.isBlank) {
```
Comment (marc0der) [inline src/main/scala/io/sdkman/vendor/release/routes/Validation.scala:71]: This validation feels a bit naive, not sure if there is anything more that we can do. Would some regex perhaps help here?

**Comment by marc0der** [comment]
Looking good so far! I've left a bunch of comments, but I'm probably being unclear in my explanations :-) I'd be happy to pair up with you sometime if that would be easier. Alternatively, I could do the validation stuff as a subsequent commit on top of yours.

File: src/test/resources/features/release_version_checksums.feature (hunk 1 above)
Comment (hgeraldino) [reply src/test/resources/features/release_version_checksums.feature:8]: Switched to groovy (UNIVERSAL)

File: src/main/scala/io/sdkman/vendor/release/routes/Validation.scala (hunk 8 above)
Comment (hgeraldino) [reply src/main/scala/io/sdkman/vendor/release/routes/Validation.scala:71]: Ok, regex makes sense. I'll add that

**Comment by marc0der** [comment]
Great work @hgeraldino! I'm going to do one or two small tweaks but will release it soon.

**Pull request merged**